## importations des données

In [1]:
# importation des packages

import pandas as pd
import duckdb as db
import zipfile
import csv

In [2]:
# importation des données

chemin_vers_fichier_zip = 'C:/Users/Moi/sql_challenge/database/inventory_items.csv.zip'

# Ouvrir le fichier zip
with zipfile.ZipFile(chemin_vers_fichier_zip, 'r') as z:
    # Liste des fichiers contenus dans le zip
    liste_fichiers = z.namelist()
    # On suppose qu'il n'y a qu'un fichier CSV dans le zip
    nom_fichier_csv = liste_fichiers[0]
    # Lire le fichier CSV dans une dataframe
    with z.open(nom_fichier_csv) as f:
        inventory_items = pd.read_csv(f, encoding='latin1')

# Afficher les 5 premières lignes de la dataframe
inventory_items.head(2)

id  product_id                 created_at                    sold_at  \
0  67971       13844  2022-07-02 07:09:20+00:00  2022-07-24 06:33:20+00:00   
1  67972       13844  2023-12-20 03:28:00+00:00                        NaN   

      cost product_category            product_name       product_brand  \
0  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   
1  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   

   product_retail_price product_department                       product_sku  \
0                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   
1                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   

   product_distribution_center_id  
0                               7  
1                               7

In [11]:
# importation des données

chemin_vers_fichier_zip = 'C:/Users/Moi/sql_challenge/database/events.zip'

# Ouvrir le fichier zip
with zipfile.ZipFile(chemin_vers_fichier_zip, 'r') as z:
    # Liste des fichiers contenus dans le zip
    liste_fichiers = z.namelist()
    # On suppose qu'il n'y a qu'un fichier CSV dans le zip
    nom_fichier_csv = liste_fichiers[0]
    # Lire le fichier CSV dans une dataframe
    with z.open(nom_fichier_csv) as f:
        events_temp = pd.read_csv(f, encoding='latin1')

# Afficher les 5 premières lignes de la dataframe
events_temp.head(2)

C:\Users\Moi\AppData\Local\Temp\ipykernel_3396\3458977689.py:13: DtypeWarning: Columns (13) have mixed types. Specify dtype option on import or set low_memory=False.
  events_temp = pd.read_csv(f, encoding='latin1')


id  user_id  sequence_number                            session_id  \
0  2198523      NaN                3  83889ed2-2adc-4b9a-af5d-154f6998e778   
1  1773216      NaN                3  7a3fc3f2-e84f-44fe-8876-eff76741f7a3   

                  created_at     ip_address          city      state  \
0  2021-06-17 17:30:00+00:00  138.143.9.202     SÆo Paulo  SÆo Paulo   
1  2020-08-07 08:41:00+00:00  85.114.141.79  Santa Isabel  SÆo Paulo   

  postal_code browser traffic_source      uri event_type Unnamed: 13  
0   02675-031  Chrome        Adwords  /cancel     cancel         NaN  
1   07500-000  Safari        Adwords  /cancel     cancel         NaN

In [12]:
events = events_temp.copy()

In [30]:
events = events.astype({col: str for col in events.select_dtypes(include='object').columns})

In [31]:
events.head(1)

id  user_id  sequence_number                            session_id  \
0  2198523      NaN                3  83889ed2-2adc-4b9a-af5d-154f6998e778   

                  created_at     ip_address       city      state postal_code  \
0  2021-06-17 17:30:00+00:00  138.143.9.202  SÆo Paulo  SÆo Paulo   02675-031   

  browser traffic_source      uri event_type Unnamed: 13  
0  Chrome        Adwords  /cancel     cancel         nan

In [35]:
events.dtypes

id                   int64
user_id            float64
sequence_number      int64
session_id          object
created_at          object
ip_address          object
city                object
state               object
postal_code         object
browser             object
traffic_source      object
uri                 object
event_type          object
Unnamed: 13         object
dtype: object

In [4]:
users = pd.read_csv('C:/Users/Moi/sql_challenge/database/users.csv')
users.head(3)

id   first_name last_name                           email  age gender  \
0    457      Timothy      Bush         timothybush@example.net   65      M   
1   6578    Elizabeth  Martinez   elizabethmartinez@example.com   34      F   
2  36280  Christopher   Mendoza  christophermendoza@example.net   13      M   

  state            street_address postal_code        city country  latitude  \
0  Acre       87620 Johnson Hills   69917-400  Rio Branco  Brasil -9.945568   
1  Acre         1705 Nielsen Land   69917-400  Rio Branco  Brasil -9.945568   
2  Acre  125 Turner Isle Apt. 264   69917-400  Rio Branco  Brasil -9.945568   

   longitude traffic_source                 created_at  
0  -67.83561         Search  2022-07-19 13:51:00+00:00  
1  -67.83561         Search  2023-11-08 18:49:00+00:00  
2  -67.83561          Email  2019-08-24 06:10:00+00:00

In [5]:
products = pd.read_csv('C:/Users/Moi/sql_challenge/database/products.csv')
products.head(3)

id     cost     category  \
0  13842  2.51875  Accessories   
1  13928  2.33835  Accessories   
2  14115  4.87956  Accessories   

                                                name brand  retail_price  \
0   Low Profile Dyed Cotton Twill Cap - Navy W39S55D    MG          6.25   
1  Low Profile Dyed Cotton Twill Cap - Putty W39S55D    MG          5.95   
2       Enzyme Regular Solid Army Caps-Black W35S45D    MG         10.99   

  department                               sku  distribution_center_id  
0      Women  EBD58B8A3F1D72F4206201DA62FB1204                       1  
1      Women  2EAC42424D12436BDD6A5B8A88480CC3                       1  
2      Women  EE364229B2791D1EF9355708EFF0BA34                       1

In [6]:
order_items = pd.read_csv('C:/Users/Moi/sql_challenge/database/order_items.csv')
order_items.head(3)

id  order_id  user_id  product_id  inventory_item_id     status  \
0  152013    104663    83582       14235             410368  Cancelled   
1   40993     28204    22551       14235             110590   Complete   
2   51224     35223    28215       14235             138236   Complete   

                  created_at                 shipped_at  \
0  2023-05-07 06:08:40+00:00                        NaN   
1  2023-03-14 03:47:21+00:00  2023-03-15 22:57:00+00:00   
2  2023-12-05 13:25:30+00:00  2023-12-06 01:20:00+00:00   

                delivered_at returned_at  sale_price  
0                        NaN         NaN        0.02  
1  2023-03-18 01:08:00+00:00         NaN        0.02  
2  2023-12-10 10:04:00+00:00         NaN        0.02

In [7]:
orders = pd.read_csv('C:/Users/Moi/sql_challenge/database/orders.csv')
orders.head(3)

order_id  user_id     status gender                 created_at returned_at  \
0         8        5  Cancelled      F  2022-10-20 10:03:00+00:00         NaN   
1        60       44  Cancelled      F  2023-01-20 02:12:00+00:00         NaN   
2        64       46  Cancelled      F  2021-12-06 09:11:00+00:00         NaN   

  shipped_at delivered_at  num_of_item  
0        NaN          NaN            3  
1        NaN          NaN            1  
2        NaN          NaN            1

In [8]:
distribution_centers = pd.read_csv('C:/Users/Moi/sql_challenge/database/distribution_centers.csv')
distribution_centers.head(3)

id        name  latitude  longitude
0   1  Memphis TN   35.1174   -89.9711
1   2  Chicago IL   41.8369   -87.6847
2   3  Houston TX   29.7604   -95.3698

## Analyse des sources de trafic

In [49]:
query = """ 
SELECT id, user_id, sequence_number, created_at, ip_address, browser, traffic_source, event_type
FROM events
WHERE session_id ='45ce8e0a-1659-4039-82fb-356f9848e60d'
"""

db.sql(query)

┌─────────┬─────────┬─────────────────┬────────────────────────┬───────────────┬─────────┬────────────────┬────────────┐
│   id    │ user_id │ sequence_number │       created_at       │  ip_address   │ browser │ traffic_source │ event_type │
│  int64  │ double  │      int64      │        varchar         │    varchar    │ varchar │    varchar     │  varchar   │
├─────────┼─────────┼─────────────────┼────────────────────────┼───────────────┼─────────┼────────────────┼────────────┤
│ 1046912 │ 80041.0 │              13 │ 2023-06-24 16:13:34+…  │ 130.75.147.39 │ Chrome  │ Email          │ purchase   │
│ 1046910 │ 80041.0 │              11 │ 2023-06-21 16:11:04+…  │ 130.75.147.39 │ Chrome  │ Email          │ product    │
│ 1046900 │ 80041.0 │               1 │ 2023-06-21 15:59:22+…  │ 130.75.147.39 │ Chrome  │ Email          │ department │
│ 1046905 │ 80041.0 │               6 │ 2023-06-21 16:03:17+…  │ 130.75.147.39 │ Chrome  │ Email          │ cart       │
│ 1046904 │ 80041.0 │           

nous avons recuperer une session, et nous constatons bien evidement qu'il corespond à un seul et unique utilisateur '80041'. cet utilisateur 
a effectuer plusieurs evenement ou interaction dont la première est le clique sur une catégorie de produits (department), puis l'ajout au panier et 
à la fin (13 : sequence_number) il a acheté le produit avant de quiter sa session.
On constate aussi que l'action d'achat a été effectué le 24/06/2024 dont 3 jours après avoir consulter le produit et le mettre dans le panier.
donc ce qui voudrais dire que l'orsqu'un user ouvre une session, il peut réactiver la meme session et poursuivre son parcours d'achat.

Voyons voir si ce client peut avoir plusieurs session à la fois.

In [51]:
query = """ 
SELECT id, session_id, sequence_number, created_at, ip_address, browser, traffic_source, event_type
FROM events
WHERE user_id =80041.0
"""

db.sql(query)

┌─────────┬──────────────────────┬─────────────────┬──────────────────────┬───┬─────────┬────────────────┬────────────┐
│   id    │      session_id      │ sequence_number │      created_at      │ … │ browser │ traffic_source │ event_type │
│  int64  │       varchar        │      int64      │       varchar        │   │ varchar │    varchar     │  varchar   │
├─────────┼──────────────────────┼─────────────────┼──────────────────────┼───┼─────────┼────────────────┼────────────┤
│ 1046912 │ 45ce8e0a-1659-4039…  │              13 │ 2023-06-24 16:13:3…  │ … │ Chrome  │ Email          │ purchase   │
│ 1046883 │ 37479389-5e6f-4a55…  │              10 │ 2023-06-21 16:25:4…  │ … │ Firefox │ Adwords        │ department │
│ 1046892 │ 088032a4-2d5b-4040…  │               6 │ 2023-06-21 14:25:3…  │ … │ Chrome  │ Email          │ cart       │
│ 1046877 │ 37479389-5e6f-4a55…  │               4 │ 2023-06-21 16:16:3…  │ … │ Firefox │ Adwords        │ department │
│ 1046922 │ 7f1b4e9e-2e18-40cf…  │      

In [55]:
query = """ 
SELECT order_id,status,gender, created_at, returned_at, shipped_at, delivered_at
FROM orders
WHERE user_id =80041.0
"""

db.sql(query)

┌──────────┬───────────┬─────────┬───────────────────────────┬─────────────┬────────────┬──────────────┐
│ order_id │  status   │ gender  │        created_at         │ returned_at │ shipped_at │ delivered_at │
│  int64   │  varchar  │ varchar │          varchar          │   varchar   │  varchar   │   varchar    │
├──────────┼───────────┼─────────┼───────────────────────────┼─────────────┼────────────┼──────────────┤
│   100264 │ Cancelled │ F       │ 2023-06-21 16:24:00+00:00 │ NULL        │ NULL       │ NULL         │
└──────────┴───────────┴─────────┴───────────────────────────┴─────────────┴────────────┴──────────────┘

In [54]:
query = """ 
SELECT DISTINCT status 
FROM orders
"""

db.sql(query)

┌────────────┐
│   status   │
│  varchar   │
├────────────┤
│ Shipped    │
│ Processing │
│ Cancelled  │
│ Complete   │
│ Returned   │
└────────────┘

In [56]:
events.dtypes

id                   int64
user_id            float64
sequence_number      int64
session_id          object
created_at          object
ip_address          object
city                object
state               object
postal_code         object
browser             object
traffic_source      object
uri                 object
event_type          object
Unnamed: 13         object
dtype: object

### QUESTION 1 :  D’où proviennent la majorité des sessions de site web ? Comment elles sont reparties par source de campagne, par source de navigateur ?

In [60]:
query = """ 
SELECT traffic_source, 
       COUNT(DISTINCT session_id) AS num_of_sessions
FROM events
GROUP BY traffic_source
         
ORDER BY num_of_sessions DESC
"""

db.sql(query)

┌────────────────┬─────────────────┐
│ traffic_source │ num_of_sessions │
│    varchar     │      int64      │
├────────────────┼─────────────────┤
│ Email          │          232431 │
│ Adwords        │          155550 │
│ YouTube        │           51865 │
│ Facebook       │           51548 │
│ Organic        │           26151 │
└────────────────┴─────────────────┘

Les campagnes par email sont la source de trafic la plus importante, représentant le plus grand nombre de sessions.
Analyse : Cela peut indiquer une base d'abonnés solide et des campagnes d'emailing efficaces. Il pourrait être intéressant de continuer à investir dans l'email marketing en optimisant les contenus et les envois pour maintenir ou augmenter cet engagement.

Les annonces payantes via Google Adwords sont la deuxième source de trafic.
Analyse : Adwords joue un rôle crucial dans la génération de trafic payant. L'investissement dans ces annonces semble être rentable, et il pourrait être utile d'analyser les campagnes spécifiques pour voir lesquelles sont les plus performantes et pourquoi.

Cette analyse des sources de trafic montre que les emails et les annonces payantes via Google Adwords sont les plus efficaces pour générer des sessions. Toutefois, il est important de diversifier les sources de trafic et d'optimiser chaque canal pour maximiser le retour sur investissement et atteindre un public plus large.

In [62]:
query = """ 
SELECT traffic_source,
       browser,
       COUNT(DISTINCT session_id) AS num_of_sessions
FROM events
GROUP BY traffic_source,
         browser
         
ORDER BY num_of_sessions DESC
"""

db.sql(query)

┌────────────────┬─────────┬─────────────────┐
│ traffic_source │ browser │ num_of_sessions │
│    varchar     │ varchar │      int64      │
├────────────────┼─────────┼─────────────────┤
│ Email          │ Chrome  │          116232 │
│ Adwords        │ Chrome  │           77880 │
│ Email          │ Firefox │           46722 │
│ Email          │ Safari  │           46217 │
│ Adwords        │ Firefox │           31156 │
│ Adwords        │ Safari  │           30966 │
│ YouTube        │ Chrome  │           25896 │
│ Facebook       │ Chrome  │           25649 │
│ Organic        │ Chrome  │           13227 │
│ Email          │ IE      │           11714 │
│   ·            │ ·       │              ·  │
│   ·            │ ·       │              ·  │
│   ·            │ ·       │              ·  │
│ Adwords        │ Other   │            7811 │
│ Adwords        │ IE      │            7737 │
│ Organic        │ Safari  │            5220 │
│ Organic        │ Firefox │            5085 │
│ YouTube    

La majorité des sessions générées par les emails proviennent des utilisateurs de Chrome, ce qui est conforme aux tendances générales du marché où Chrome est le navigateur le plus utilisé.
Firefox  et Safari : Un nombre significatif de sessions provient également de ces navigateurs, ce qui montre une diversité dans la base d'utilisateurs.
 Bien que IE soit un navigateur moins populaire, il génère encore un nombre notable de sessions via les emails.

### QUESTION 2 : comprendre les taux de conversion des sessions en commandes
 D’après les résultats de la première question, les campagnes emailing et les annonces googles adwords sont les deux sources de traffic les plus importantes. 
Nous souhaitons creuser profondément ces campagnes et essayer de comprendre les taux de conversion des sessions en commandes. En d'autres termes, il s'agit de déterminer quel pourcentage des sessions se convertissent en vente pour l'entreprise.


In [66]:
query = """ 
WITH events_Email_Adwords AS (
    SELECT *
    FROM events
    WHERE traffic_source IN ('Email', 'Adwords')
)
SELECT events_Email_Adwords.traffic_source,
       COUNT(DISTINCT events_Email_Adwords.session_id) AS sessions,
       COUNT(DISTINCT orders.order_id) AS num_of_order,
       COUNT(DISTINCT orders.order_id)/ COUNT(DISTINCT events_Email_Adwords.session_id) AS session_to_order_conv_rt
FROM events_Email_Adwords 
LEFT JOIN orders ON events_Email_Adwords.user_id = orders.user_id
GROUP BY events_Email_Adwords.traffic_source
"""

db.sql(query)

┌────────────────┬──────────┬──────────────┬──────────────────────────┐
│ traffic_source │ sessions │ num_of_order │ session_to_order_conv_rt │
│    varchar     │  int64   │    int64     │          double          │
├────────────────┼──────────┼──────────────┼──────────────────────────┤
│ Adwords        │   155550 │        70654 │       0.4542205078752813 │
│ Email          │   232431 │        89291 │      0.38416132099418754 │
└────────────────┴──────────┴──────────────┴──────────────────────────┘